### PR TITLE
Remove the special case handling of LOGICAL and i1.

### PR DESCRIPTION
**Pull request by schweitzpgi**

### CODE REVIEW

File: flang/lib/Lower/ConvertExpr.cpp [hunk 1]
```diff
@@ -218,10 +229,18 @@ class ExprLowering {
 
   mlir::Value gendef(Fortran::semantics::SymbolRef sym) { return gen(sym); }
 
+  /// Loading a LOGICAL immediately converts to an `i1` bool value.
+  mlir::Value genLoad(mlir::Value addr) {
```
Comment (vdonaldson) [inline flang/lib/Lower/ConvertExpr.cpp:233]: Is this necessary, or at least harmless in all cases?  For example, old code may allow interoperability between logical and integer objects.

File: flang/lib/Lower/Bridge.cpp [hunk 2]
```diff
@@ -71,14 +71,15 @@ struct IncrementLoopInfo {
...
-  bool isStructured() const { return headerBlock == nullptr; }
+  bool isStructured() const { return !headerBlock; }
 
   // Data members for both structured and unstructured loops.
   Fortran::semantics::Symbol *loopVariableSym;
   const Fortran::parser::ScalarExpr &lowerExpr;
   const Fortran::parser::ScalarExpr &upperExpr;
   const std::optional<Fortran::parser::ScalarExpr> &stepExpr;
   mlir::Type loopVariableType;
+
```
Comment (vdonaldson) [inline flang/lib/Lower/Bridge.cpp:82]: Is this change intentional?

Comment (schweitzpgi) [reply flang/lib/Lower/Bridge.cpp:82]: I don't recall doing it. Hmm.

File: flang/lib/Lower/ConvertExpr.cpp (hunk 1 above)
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:233]: Strictly speaking, it is not necessary. It is harmless. It simplifies internal lower of bool expressions.

There is some subtle interplay between what I'll term a "visible" LOGICAL (Fortran sense) and an "internal" bool value (semantics sense) here. In the first case, we want to expose the Fortran type since it is plainly important to memory access size, calling conventions, etc. In the latter case, we want a single type (`i1`) to use the standard ops, etc. It's a special type because legal truncation and extension have no effect on the domain of possible values.

Comment (jeanPerier) [reply flang/lib/Lower/ConvertExpr.cpp:233]: I agree with Val here. I am wary of this logical handling, flang will have different behavior around logical:
e.g:
With this change are we sure that:

```
subroutine foo(a, b)
  logical(4) :: a, b
  a = b
subroutine
```

Will preserve the physical representation ? Sure, Fortran standard does not say so, but in practice all other compilers do.

Anyway we have discussed this enough, let's keep the easy and backward incompatible way  let's wait for issues to come...

Comment (sscalpone) [reply flang/lib/Lower/ConvertExpr.cpp:233]: We had a long conversation a while back about the difference between logical and boolean.  We decide that a boolean arises from logical operation; otherwise, logicals are represented as their value.  I don't recall any effect of this rule other than loads, stores, and assignments would preserve the bit-pattern of the logical.  Explicit conversions from boolean to logical would follow a conventions, which escapes me -- was it configurable to 0 and 1 or ~0?  Conversion from a smaller to large logical would follow a similar convention to sign extend or not.

Comment (sscalpone) [reply flang/lib/Lower/ConvertExpr.cpp:233]: https://docs.google.com/document/d/1B055cThpOsJdrC9iat1j-bKgWGOdE2S3cDfW5MIH3ME/edit?usp=sharing

Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:233]: > I agree with Val here. I am wary of this logical handling, flang will have different behavior around logical:

In what regard?

> e.g:
> With this change are we sure that:
> 
> ```
> subroutine foo(a, b)
>   logical(4) :: a, b
>   a = b
> subroutine
> ```
> 
> Will preserve the physical representation ?

I fully expect that this will be lowered to

```mlir
  %load = fir.load %a : !fir.ref<!fir.logical<4>>
  ...
  fir.store %load to %b : !fir.ref<!fir.logical<4>>
```

Why? Because both A and B are observable effects, which should be preserved.

Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:233]: BTW, here is what the compiler is, in fact, doing right now.
```Fortran
subroutine s(a,b)
  LOGICAL :: a, b
  b = a
end subroutine s
```
________________

```mlir
func @_QPs(%arg0: !fir.ref<!fir.logical<4>>, %arg1: !fir.ref<!fir.logical<4>>) {
  %0 = fir.load %arg0 : !fir.ref<!fir.logical<4>>
  fir.store %0 to %arg1 : !fir.ref<!fir.logical<4>>
  return
}
```

Comment (jeanPerier) [reply flang/lib/Lower/ConvertExpr.cpp:233]: OK, sorry for writing this comment without testing. I thought it would not work but actually it works thanks to the previous change in PR #11. If it's OK to rely on FIR folding to achieve the semantics behavior, then why not.
I still think it is not a defensive approach semantically speaking and there are more complex cases where pgfortran/gfortran do preserve the layout and flang now wouldn't. For instance:  `a = MERGE(a, b, c)` now behaves differently because FIR folding does not eliminate the conversion to i1 here since fir.select is in the middle.

Anyway, I am not against changing the way it was done, I just think this change is incomplete (Regression around user and intrinsic function call interface, and MERGE behaving differently), but it's in big part my bad since we should write regression tests once and for all for the agreed cases where.
I have started to write some and will push them so we have better requirements here.

Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:233]: Honestly, we don't have to be bitwise compatible with some hypothetical Fortran implementation where .TRUE. is a constant of 37 or whatever. It's a non-issue.

We do have to be consistent with our codegen, which is LLVM. There, true is `1:i1` and false is `0:i1`. We can do it any way we want so long as we follow that rule. (Which is to say, there is no choice.)

Observable values are nothing new either. A `bool` in clang is internally `x:i1`. It is typically stored in a byte in memory.

We don't need to make it any more complicated.

When storing a _computed_ true or false value, we will store that value in some well-defined way and load it and truncate it for internal use.

If intrinsics are "fully bit preserving" (memcpy like operations), and we must preserve garbage bits, then those intrinsics must be implemented that way (as some sort of bitblt op). I don't see how that dictates the internal representation of what is true and false.
